libtransmission/peer-mgr.cc

```diff
@@ -862,89 +849,75 @@ class tr_swarm
-            TR_ASSERT(it_that->first == info_that.listen_socket_address());
-            TR_ASSERT(it_that->first.address() == info_this.listen_address());
-            TR_ASSERT(it_that->first.port() != info_this.listen_port());
+            auto const info_that = it_that->second;
+            TR_ASSERT(it_that->first == info_that->listen_socket_address());
+            TR_ASSERT(it_that->first.address() == info_this->listen_address());
+            TR_ASSERT(it_that->first.port() != info_this->listen_port());
 
-            // If there is an existing connection to this peer, keep the better one
-            if (info_that.is_connected() && on_got_port_duplicate_connection(msgs, it_that, was_connectable))
+            // if there is an existing connection to this peer, keep the better one
+            if (info_that->is_connected() && on_got_port_duplicate_connection(msgs, info_that))
             {
-                return;
+                goto EXIT; // NOLINT cppcoreguidelines-avoid-goto
             }
 
-            info_this.merge(info_that);
-            auto from = info_that.from_first();
-            stats.known_peer_from_count[from] -= connectable_pool.erase(info_that.listen_socket_address());
-        }
-        else if (!was_connectable)
-        {
-            info_this.set_connectable();
-        }
+            // merge the peer info objects
+            info_this->merge(*info_that);
 
-        auto nh = was_connectable ? connectable_pool.extract(info_this.listen_socket_address()) :
-                                    incoming_pool.extract(msgs->socket_address());
-        TR_ASSERT(!std::empty(nh));
-        if (was_connectable)
-        {
-            TR_ASSERT(nh.key() == nh.mapped().listen_socket_address());
+            // info_that will be replaced by info_this later, so decrement stat
+            --stats.known_peer_from_count[info_that->from_first()];
         }
-        else
+        // we are going to insert a brand-new peer info object to the pool
+        else if (std::empty(info_this->listen_port()))
         {
-            ++stats.known_peer_from_count[nh.mapped().from_first()];
-            TR_ASSERT(nh.key().address() == nh.mapped().listen_address());
+            info_this->set_connectable();
         }
-        nh.key().port_ = event.port;
-        [[maybe_unused]] auto const inserted = connectable_pool.insert(std::move(nh)).inserted;
-        TR_ASSERT(inserted);
-        info_this.set_listen_port(event.port);
 
+        // erase the old peer info entry
+        stats.known_peer_from_count[info_this->from_first()] -= connectable_pool.erase(info_this->listen_socket_address());
+
+        // set new listen port
+        info_this->set_listen_port(event.port);
+
+        // insert or replace the peer info ptr at the target location
+        ++stats.known_peer_from_count[info_this->from_first()];
+        connectable_pool.insert_or_assign(info_this->listen_socket_address(), std::move(info_this));
```

Comment:
As a thread sensitive operation, let's merge the two stats operations?
(I'm scoping the lifetime of `erased` into a block due to the `goto`)

```suggestion
        {
            // erase the old peer info entry
            auto const erased = connectable_pool.erase(info_this->listen_socket_address());

            // set new listen port
            info_this->set_listen_port(event.port);

            // insert or replace the peer info ptr at the target location
            connectable_pool.insert_or_assign(info_this->listen_socket_address(), std::move(info_this));

            // update stats
            stats.known_peer_from_count[info_this->from_first()] += 1 - erased;
        }
```

Reply:
I'm against this. I think this is not worth the less intuitive code, and this part is complicated enough as-is.

Either way, the stat operations here won't be atomic, so this does not eliminate the race condition. It at most reduces the chance of the race condition doing any harm, by moving the decrement and increment operations closer together in time.

Reply:
OK. Not a blocker.